Translucency Sorting With Quad Splitting

douira opened the pull request:
- Fixes translucency sorting for geometry that is intersecting or otherwise unsortable by splitting quads in the partition tree.
- Introduces a setting with the states "off", "safe", "unlimited" that controls to which degree quads are split. The default is "safe", where it doesn't disable normal quantization and limits how many quads can be generated compared to the original number. (i.e. the geometry amplification is capped)
- Removes some unnecessary classes from the translucent data class hierarchy
- Simplifies vertex/quad count handling and translucent data object construction
- Reorders translucent sorting to be before meshing so that modifications to meshing can be performed
- Refactors `TQuad` construction to happen mostly in its own class

A 1.21.1 version of this patch is available at https://github.com/douira/sodium/tree/1.21.1/translucent-quad-splitting (currently outdated)

Fixes https://github.com/CaffeineMC/sodium/issues/2943

![overview](https://github.com/user-attachments/assets/59f55e59-b61f-4567-9601-a16635cb839f)
![split](https://github.com/user-attachments/assets/892f4b08-f9d8-498a-90b1-92b00f9efa2d)

douira committed:
refactor old translucent data reuse to be a method with implementations instead of using instanceof

Signed-off-by: douira <douira100@gmail.com>

douira committed:
move quad analysis code out of TranslucentGeometryCollector and fix some bugs around quad count handling and quad init

douira committed:
fix bsp triggering with negative, fix quad appending logic, fix quad update logic, fix split case logic to take on-plane vertices into account

douira committed:
prefer splitting on the y-axis, extract splitting group update into a method

douira committed:
implement corner-clipping quad splitting and various triangle-related (degenerate quad) cases

douira commented:
It does break things that require sort triggering on unaligned normals but it's unlikely anyone would notice unless the geometry is particularly weird. It only becomes more noticeable with this patch because previously the partition tree wouldn't generate unaligned partition (i.e. triggering) planes, and unaligned trigger planes were only generated by the super rarely used dynamic topo sorting mode.

I could split it out if you'd merge such a fix before 0.7, which is when I'd expect this to be ready for as a whole.

muzikbike commented:
Does this fix #2943? May be worth marking as such if so.

douira committed:
add missing quad extent updates for triangle splits

douira committed:
simplify regular and full quad init procedure

douira committed:
add quad splitting mode setting and integrate it with the partition tree builder and the topo sorter

douira committed:
fix wrong quad splitting factor interpretation when it's infinite

douira committed:
use epsilon for determining whether vertices are on the split plane to avoid getting into an inconsistent when later calculations yield different results because precision was lost along the way

douira committed:
improve label of quad splitting option, make it disabled if  translucency sorting is disabled overall

douira committed:
fix a bug with the plane-based triggering system where sections added later would not be added to the triggering system even though they should have been, this was caused by a buggy equals implementation in AlignableNormal.
remove AlignableNormal and replace with a Vector3f & int pair in NormalList and NormalPlanes, associated refactors to remove custom hashing strategy usage again,

douira committed:
don't write quads to the update buffer multiple times even when they get updated repeatedly

muzikbike commented:
Does this also fix #2802? Images posted in the discord thread certainly appear to imply so.

douira committed:
more quad splitting wip

douira committed:
cleanup unused code

douira committed:
move updated vertex count calculation into separate method

douira committed:
fix a case of triangle splitting wrongly deciding which side to move an on-plane vertex to

douira committed:
use the correctly cleaned normal for initializing the unaligned normal planes set

douira committed:
display splitting mode in the TS-related debug line

douira committed:
fix splitting for triangles where at least one vertex is on the split plane

douira committed:
fix quantization of split plane normals

douira committed:
fix some splitting errors caused by misclassification of quads with two nearly identical vertices

douira committed:
filter out invalid quads even if they are invalid before being processed by the bsp tree

douira committed:
Completely refactor the mesh modification system to correctly distinguish between index and mesh sizes.
Revert ChunkMeshBufferBuilder to be mostly how it was before, without write methods and instead allow it to write to an external buffer.
Modified quads are no longer repeatedly written during splitting and instead just written out once in their modified form.
Clarified the difference between index and mesh quad counts throughout the translucency sorting code.
Removed MixedDirectionData and SplitDirectionData as part of moving the translucency sorting before the meshing code so that meshes may be modified more easily.

douira committed:
remove ensureDirectionsAre mixed assertion that is no longer necessary

douira committed:
use epsilon to prevent vanishing split distances from causing a division by zero assertion to fail

douira committed:
make same vertex detection use an epsilon too

douira committed:
improve quantized normal to facing calculation

douira committed:
fix rare bug that caused non-coplanar quads to be calculated to be in a shared splitting group

douira committed:
disable normal quantization for trigger planes in unlimited mode for perfect triggering at the cost of many more normal planes being used.

douira committed:
prune diff in preparation for bringing the pr out of draft mode

douira committed:
remove quad splitting debug info

douira commented:
This is ready for review and merging. No open in-scope issues found during testing remain.

Note that the bug fix PR https://github.com/CaffeineMC/sodium/pull/3033 is included in this for testing reasons and once it gets merged into upstream I'll drop it from this.
